ILASM - Fixing ILASM roundtrip issue with `vararg` calling convention

TIHan opened the pull request:
Should resolve: https://github.com/dotnet/runtime/issues/72759

**Description**

ILASM was failing to emit the correct callsite signature for a method with a `vararg` calling convention if one of the parameters was generic instantiation of a type.

In `ResolveLocalMemberRefs`, it seems we lost some information when getting the fixed signature for `vararg` callconv and it creates a bad signature for the new member reference.

There are two solutions:

1. Remove special handling for `vararg` callconv in `ResolveLocalMemberRefs` as we already have the correct signature from `m_LocalMethodRefDList`.
2. Fix `CountBytesOfOneArg` to handle `ELEMENT_TYPE_GENERICINST` as described by @jakobbotsch 

The PR is now using solution number 2.

I manually tested this and it resolves the issue reported. I also tried this scenario:
ClassLibrary1.dll
```csharp
using System.Runtime.Intrinsics;

namespace ClassLibrary1
{
    public class Class1
    {
        public static int paramLength(Vector128<int> arg, __arglist)
        {
            ArgIterator iterator = new ArgIterator(__arglist);
            return iterator.GetRemainingCount();
        }
    }
}
```
vararg_test.exe
```csharp
using ClassLibrary1;
using System.Runtime.Intrinsics;

Console.WriteLine(Class1.paramLength(Vector128<int>.AllBitsSet, __arglist(1, 2)));
```
This scenario works with the PR changes, so this gives me confidence in the change as this is across assembly boundaries.

**Acceptance Criteria**
- [x] No additional ilasm roundtrip failures in CI

TIHan commented:
@dotnet/jit-contrib This is ready, pending CI runs

jakobbotsch commented:
> it seems we lost some information when getting the fixed signature and it creates a bad signature for the new member reference.

There are other uses of this `_ GetFixedSigOfVarArg` function, does it need to be fixed instead it if is returning a bad signature?

If I understand the fix here right it will cause ilasm to unnecessarily create `TypeRef` tokens to types defined within the same module, which seems a little wasteful (though given that it is only for varargs, probably not a big deal).

jakobbotsch commented:
> There are other uses of this `_GetFixedSigOfVarArg` function, does it need to be fixed instead it if is returning a bad signature?

It looks suspect to me that `_CountBytesOfOneArg` used by `_GetFixedSigOfVarArg` does not handle `ELEMENT_TYPE_GENERICINST`.

TIHan commented:
> There are other uses of this _GetFixedSigOfVarArg function, does it need to be fixed instead it if is returning a bad signature?

Possibly, I didn't know exactly what might need to be handled in there and I think the runtime uses this as well. To minimize any risk, I went with just the change only in ILASM.

> If I understand the fix here right it will cause ilasm to unnecessarily create TypeRef

Is this when the `tkMemberDef` is 0? When debugging, it actually does not go down that path.

> It looks suspect to me that _CountBytesOfOneArg used by _GetFixedSigOfVarArg does not handle ELEMENT_TYPE_GENERICINST.

I think this might be it. I just tried to handle this, and when checking it under ILDASM I can see the parameter type starting to form:
```
    IL_000b:  call       vararg int32 Runtime_71375::VarArgs(int32,
                                                             int32,
                                                             int32,
                                                             int32,
                                                             int32,
                                                             int32,
                                                             valuetype [System.Runtime.Intrinsics]System.Runtime.Intrinsics.Vector128`1<void,,int32,,,,,>)
```
I'll see if I can get it working.

TIHan committed:
Revert change in ILASM in favor of handling the fix in CountBytesOfOneArg

TIHan commented:
Updated the PR to handle `ELEMENT_TYPE_GENERICINST` in `CountBytesOfOneArg` and reverted the other fix in favor of it.

@dotnet/jit-contrib ready for review again pending CI runs.

jkotas commented on src/coreclr/inc/corhlprpriv.cpp:
```diff
@@ -109,6 +109,23 @@ HRESULT _CountBytesOfOneArg(
     switch (ulElementType)
     {
+        case ELEMENT_TYPE_GENERICINST:
+            CHECK_REMAINDER;
```
Should `CHECK_REMAINDER` be in front of `CorSigUncompressData` instead?

jkotas replied:
And there should be `cb = cbTotalMax - cbTotal;` in front of `_CountBytesOfOneArg` calls.

TIHan replied:
I see, will fix that.

TIHan committed:
feedback

jakobbotsch commented on src/coreclr/inc/corhlprpriv.cpp:
```diff
@@ -109,6 +109,27 @@ HRESULT _CountBytesOfOneArg(
     {
+        case ELEMENT_TYPE_GENERICINST:
+            // skip over generic type
+            CHECK_REMAINDER;
```
I guess this is not strictly necessary in front of the `_CountBytesOfOneArg` calls, since the function ensures this is called before it reads anything from the signature.

jakobbotsch replied:
Actually I see the other cases do the same thing, so nevermind this.

TIHan replied:
Sure, but looking at places where we call `_CountBytesOfOneArg` in here, it has a `CHECK_REMAINDER` before it.
